### PR TITLE
`sage.crypto`: Update `# needs`, modularization fixes

### PR DESCRIPTION
**Pull request by mkoeppe**
- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/crypto/classical_cipher.py [hunk 1]
```diff
@@ -165,25 +170,23 @@ def __init__(self, parent, key):
         EXAMPLES::
 
             sage: S = AlphabeticStrings()
-            sage: E = HillCryptosystem(S,3)
-            sage: E
+            sage: E = HillCryptosystem(S,3); E                                          # needs sage.modules
```
Comment (dcoudert) [inline src/sage/crypto/classical_cipher.py]: block scope

File: src/sage/crypto/lattice.py [hunk 2]
```diff
@@ -35,22 +36,24 @@ def gen_lattice(type='modular', n=4, m=8, q=11, seed=None,
...
+        of lattice used for testing basis reduction algorithms
+        proposed by Goldstein and Mayer [GM2002]_.
+      - ``'ideal'`` -- Special case of modular. Allows for a more
+        compact representation proposed by [LM2006]_.
+      - ``'cyclotomic'`` -- Special case of ideal. Allows for
+        efficient processing proposed by [LM2006]_.
+
+    - ``n`` -- Determinant size, primal: `det(L) = q^n`, dual: `det(L) = q^{m-n}`.
       For ideal lattices this is also the degree of the quotient polynomial.
     - ``m`` -- Lattice dimension, `L \subseteq Z^m`.
```
Comment (dcoudert) [inline src/sage/crypto/lattice.py:54]: you inserted empty lines lines 39 and 50 to separate items but not here. Shouldn't it be the same for all items of the list ?

File: src/sage/crypto/block_cipher/des.py [hunk 3]
```diff
@@ -511,7 +512,7 @@ def encrypt(self, plaintext, key):
             sage: des.encrypt(P, K56) == C
             True
         """
-        if isinstance(plaintext, (list, tuple, Vector_mod2_dense)):
+        if isinstance(plaintext, (list, tuple, Vector)):
```
Comment (dcoudert) [inline src/sage/crypto/block_cipher/des.py:515]: are we sure that this change is safe for users ? I'm not sure it's currently tested...

**Review by dcoudert — CHANGES_REQUESTED** [review]

File: src/sage/crypto/block_cipher/des.py (hunk 3 above)
Comment (mkoeppe) [reply src/sage/crypto/block_cipher/des.py:515]: Yes, this just distinguishes scalar input from list/vector input. In fact, when neither of the two conditions holds, `inputType` remains uninitialized, leading to an error in the last line of the function

File: src/sage/crypto/lattice.py (hunk 2 above)
Comment (mkoeppe) [reply src/sage/crypto/lattice.py:54]: OK, done in 2f0f2f6fb83

File: src/sage/crypto/classical_cipher.py (hunk 1 above)
Comment (mkoeppe) [reply src/sage/crypto/classical_cipher.py]: Thanks, done in 826f72dba5c

**Comment by dcoudert** [comment]
could you fix the following warnings:
```
sage -t --long --warn-long 33.3 --random-seed=250269248297457194959005008327880188134 src/sage/crypto/public_key/blum_goldwasser.py
**********************************************************************
File "src/sage/crypto/public_key/blum_goldwasser.py", line 133, in sage.crypto.public_key.blum_goldwasser.BlumGoldwasser
Warning: Variable 'q' referenced here was set only in doctest marked '# needs sage.combinat sage.symbolic'
    gcd(p, q) == a*p + b*q == 1
**********************************************************************
File "src/sage/crypto/public_key/blum_goldwasser.py", line 133, in sage.crypto.public_key.blum_goldwasser.BlumGoldwasser
Warning: Variable 'p' referenced here was set only in doctest marked '# needs sage.combinat sage.symbolic'
    gcd(p, q) == a*p + b*q == 1
**********************************************************************
File "src/sage/crypto/public_key/blum_goldwasser.py", line 133, in sage.crypto.public_key.blum_goldwasser.BlumGoldwasser
Warning: Variable 'a' referenced here was set only in doctest marked '# needs sage.combinat sage.symbolic'
    gcd(p, q) == a*p + b*q == 1
    [125 tests, 0.07 s]
sage -t --long --warn-long 33.3 --random-seed=250269248297457194959005008327880188134 src/sage/crypto/mq/sr.py
**********************************************************************
File "src/sage/crypto/mq/sr.py", line 2073, in sage.crypto.mq.sr.SR_generic.polynomial_system
Warning: Variable 's' referenced here was set only in doctest marked '# needs sage.modules sage.rings.finite_rings sage.rings.polynomial.pbori'
    F.subs(s).groebner_basis() # long time
**********************************************************************
File "src/sage/crypto/mq/sr.py", line 2073, in sage.crypto.mq.sr.SR_generic.polynomial_system
Warning: Variable 'F' referenced here was set only in doctest marked '# needs sage.modules sage.rings.finite_rings sage.rings.polynomial.pbori'
    F.subs(s).groebner_basis() # long time
    [368 tests, 10.43 s]
```

We have also several compilation warnings that should be fixed in the future

**Comment by mkoeppe** [comment]
done in 37a6b46ba75

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**